"""Fen√™tre d'aide avec documentation compl√®te YAML."""

from pathlib import Path
from PySide6.QtWidgets import QDialog
from PySide6 import QtUiTools


class HelpWindow(QDialog):
    """Fen√™tre d'aide affichant la documentation YAML compl√®te."""
    
    def __init__(self, parent=None):
        """Initialise la fen√™tre d'aide.
        
        Args:
            parent: Widget parent
        """
        super().__init__(parent)
        self._load_ui()
        self._load_stylesheet()
        self._connect_signals()
        self._populate_content()
    
    def _load_ui(self):
        """Charge l'interface depuis le fichier .ui."""
        from PySide6.QtWidgets import QTabWidget, QTextBrowser, QPushButton
        
        ui_file = Path(__file__).parent / "help_window.ui"
        loader = QtUiTools.QUiLoader()
        ui = loader.load(str(ui_file))
        
        # Copier les propri√©t√©s
        self.setWindowTitle(ui.windowTitle())
        self.resize(ui.size())
        
        # R√©cup√©rer le layout de l'UI charg√©e et l'appliquer √† ce dialog
        ui_layout = ui.layout()
        self.setLayout(ui_layout)
        
        # R√©cup√©rer les widgets
        self.tab_widget = self.findChild(QTabWidget, "tabWidget")
        self.intro_text = self.findChild(QTextBrowser, "introText")
        self.structure_text = self.findChild(QTextBrowser, "structureText")
        self.arguments_text = self.findChild(QTextBrowser, "argumentsText")
        self.shared_text = self.findChild(QTextBrowser, "sharedText")
        self.validation_text = self.findChild(QTextBrowser, "validationText")
        self.examples_text = self.findChild(QTextBrowser, "examplesText")
        self.close_button = self.findChild(QPushButton, "closeButton")
    
    def _load_stylesheet(self):
        """Charge la feuille de style depuis le fichier .qss."""
        qss_file = Path(__file__).parent / "help_window.qss"
        if qss_file.exists():
            with open(qss_file, "r", encoding="utf-8") as f:
                self.setStyleSheet(f.read())
    
    def _connect_signals(self):
        """Connecte les signaux des widgets."""
        if self.close_button:
            self.close_button.clicked.connect(self.accept)
    
    def _populate_content(self):
        """Remplit le contenu de chaque onglet."""
        self._populate_intro()
        self._populate_structure()
        self._populate_arguments()
        self._populate_shared()
        self._populate_validation()
        self._populate_examples()
    
    def _populate_intro(self):
        """Remplit l'onglet Introduction."""
        content = """
        <h2>Qu'est-ce qu'un fichier YAML de t√¢che ?</h2>
        <p>
        Un fichier YAML de t√¢che permet de d√©finir une s√©quence de commandes Windows CLI 
        qui seront ex√©cut√©es dans un ordre pr√©cis. C'est un format simple et lisible pour 
        cr√©er des workflows automatis√©s.
        </p>
        
        <h3>Avantages</h3>
        <ul>
            <li><b>Simplicit√©</b> : Format texte facile √† lire et √† modifier</li>
            <li><b>R√©utilisabilit√©</b> : D√©finissez une fois, ex√©cutez plusieurs fois</li>
            <li><b>Arguments partag√©s</b> : √âvitez la duplication avec des variables communes</li>
            <li><b>Validation</b> : V√©rification automatique des valeurs saisies</li>
            <li><b>Documentation int√©gr√©e</b> : Descriptions et noms explicites</li>
        </ul>
        
        <h3>Emplacement des fichiers</h3>
        <p>
        Les fichiers YAML doivent √™tre plac√©s dans le dossier : 
        <code>command_builder/data/tasks/</code>
        </p>
        <p>
        L'application charge automatiquement tous les fichiers <code>.yaml</code> ou <code>.yml</code> 
        pr√©sents dans ce dossier au d√©marrage.
        </p>
        
        <h3>Rechargement</h3>
        <p>
        Apr√®s avoir cr√©√© ou modifi√© un fichier YAML, red√©marrez l'application pour voir les changements.
        </p>
        """
        self.intro_text.setHtml(content)
    
    def _populate_structure(self):
        """Remplit l'onglet Structure."""
        content = """
        <h2>Structure d'un fichier YAML</h2>
        
        <h3>Champs obligatoires</h3>
        <table border="1" cellpadding="8" cellspacing="0" style="border-collapse: collapse; width: 100%;">
            <tr style="background-color: #e3f2fd;">
                <th>Champ</th>
                <th>Type</th>
                <th>Description</th>
            </tr>
            <tr>
                <td><code>name</code></td>
                <td>string</td>
                <td>Nom de la t√¢che (affich√© dans la liste)</td>
            </tr>
            <tr>
                <td><code>description</code></td>
                <td>string</td>
                <td>Description d√©taill√©e de la t√¢che</td>
            </tr>
            <tr>
                <td><code>commands</code></td>
                <td>list</td>
                <td>Liste des commandes √† ex√©cuter</td>
            </tr>
        </table>
        
        <h3>Champs optionnels</h3>
        <table border="1" cellpadding="8" cellspacing="0" style="border-collapse: collapse; width: 100%;">
            <tr style="background-color: #e3f2fd;">
                <th>Champ</th>
                <th>Type</th>
                <th>Description</th>
            </tr>
            <tr>
                <td><code>shared_arguments</code></td>
                <td>list</td>
                <td>Arguments communs √† plusieurs commandes</td>
            </tr>
        </table>
        
        <h3>Exemple minimal</h3>
        <pre style="background-color: #f5f5f5; padding: 15px; border-radius: 4px; border-left: 4px solid #2196F3;">
name: "Ma premi√®re t√¢che"
description: "Une t√¢che simple avec une commande"
commands:
  - name: "Afficher un message"
    description: "Affiche Hello World"
    command: "echo Hello World"
    arguments: []
        </pre>
        
        <h3>Structure d'une commande</h3>
        <table border="1" cellpadding="8" cellspacing="0" style="border-collapse: collapse; width: 100%;">
            <tr style="background-color: #e3f2fd;">
                <th>Champ</th>
                <th>Type</th>
                <th>Obligatoire</th>
                <th>Description</th>
            </tr>
            <tr>
                <td><code>name</code></td>
                <td>string</td>
                <td>Oui</td>
                <td>Nom de la commande</td>
            </tr>
            <tr>
                <td><code>description</code></td>
                <td>string</td>
                <td>Oui</td>
                <td>Description de la commande</td>
            </tr>
            <tr>
                <td><code>command</code></td>
                <td>string</td>
                <td>Oui</td>
                <td>Commande CLI √† ex√©cuter</td>
            </tr>
            <tr>
                <td><code>arguments</code></td>
                <td>list</td>
                <td>Oui</td>
                <td>Liste des arguments (peut √™tre vide [])</td>
            </tr>
        </table>
        """
        self.structure_text.setHtml(content)
    
    def _populate_arguments(self):
        """Remplit l'onglet Arguments."""
        content = """
        <h2>D√©finition des arguments</h2>
        
        <h3>Champs d'un argument</h3>
        <table border="1" cellpadding="8" cellspacing="0" style="border-collapse: collapse; width: 100%;">
            <tr style="background-color: #e3f2fd;">
                <th>Champ</th>
                <th>Type</th>
                <th>Obligatoire</th>
                <th>Description</th>
            </tr>
            <tr>
                <td><code>code</code></td>
                <td>string</td>
                <td>Oui</td>
                <td>Identifiant unique de l'argument (ex: DATABASE_FILE)</td>
            </tr>
            <tr>
                <td><code>name</code></td>
                <td>string</td>
                <td>Oui</td>
                <td>Nom affich√© dans l'interface (ex: "Base de donn√©es")</td>
            </tr>
            <tr>
                <td><code>required</code></td>
                <td>int</td>
                <td>Oui</td>
                <td>0 = optionnel, 1 = obligatoire</td>
            </tr>
            <tr>
                <td><code>type</code></td>
                <td>string</td>
                <td>Non</td>
                <td>"file", "folder", "text" (par d√©faut: "text")</td>
            </tr>
            <tr>
                <td><code>default</code></td>
                <td>string</td>
                <td>Non</td>
                <td>Valeur par d√©faut</td>
            </tr>
            <tr>
                <td><code>validation</code></td>
                <td>dict</td>
                <td>Non</td>
                <td>R√®gles de validation (voir onglet Validation)</td>
            </tr>
        </table>
        
        <h3>Types d'arguments</h3>
        
        <h4>1. Type "text" (par d√©faut)</h4>
        <p>Champ de saisie libre pour du texte.</p>
        <pre style="background-color: #f5f5f5; padding: 15px; border-radius: 4px; border-left: 4px solid #4CAF50;">
arguments:
  - code: "TABLE_NAME"
    name: "Nom de la table"
    required: 1
    type: "text"
    default: "ma_table"
        </pre>
        
        <h4>2. Type "file"</h4>
        <p>Champ avec bouton "Parcourir" pour s√©lectionner un fichier.</p>
        <pre style="background-color: #f5f5f5; padding: 15px; border-radius: 4px; border-left: 4px solid #4CAF50;">
arguments:
  - code: "INPUT_FILE"
    name: "Fichier d'entr√©e"
    required: 1
    type: "file"
        </pre>
        
        <h4>3. Type "folder"</h4>
        <p>Champ avec bouton "Parcourir" pour s√©lectionner un dossier.</p>
        <pre style="background-color: #f5f5f5; padding: 15px; border-radius: 4px; border-left: 4px solid #4CAF50;">
arguments:
  - code: "OUTPUT_DIR"
    name: "Dossier de sortie"
    required: 1
    type: "folder"
        </pre>
        
        <h3>Arguments obligatoires vs optionnels</h3>
        <p>
        Les arguments obligatoires (<code>required: 1</code>) sont marqu√©s d'une ast√©risque rouge <span style="color: red;">*</span> 
        dans l'interface. L'application emp√™che l'ex√©cution si un champ obligatoire est vide.
        </p>
        
        <h3>Utilisation dans la commande</h3>
        <p>
        Utilisez le <code>code</code> de l'argument entre accolades <code>{}</code> dans la commande :
        </p>
        <pre style="background-color: #f5f5f5; padding: 15px; border-radius: 4px; border-left: 4px solid #FF9800;">
command: "python script.py --input {INPUT_FILE} --output {OUTPUT_DIR}"
        </pre>
        """
        self.arguments_text.setHtml(content)
    
    def _populate_shared(self):
        """Remplit l'onglet Arguments Partag√©s."""
        content = """
        <h2>üîó Arguments partag√©s</h2>
        
        <h3>Qu'est-ce qu'un argument partag√© ?</h3>
        <p>
        Un argument partag√© est une valeur commune utilis√©e par plusieurs commandes. 
        Au lieu de d√©finir le m√™me argument pour chaque commande, vous le d√©finissez 
        une seule fois au niveau de la t√¢che.
        </p>
        
        <h3>Avantages</h3>
        <ul>
            <li>√âvite la duplication</li>
            <li>L'utilisateur saisit la valeur une seule fois</li>
            <li>Facilite la maintenance</li>
            <li>R√©duit les erreurs de saisie</li>
        </ul>
        
        <h3>D√©finition</h3>
        <p>Les arguments partag√©s se d√©finissent au niveau de la t√¢che :</p>
        <pre style="background-color: #f5f5f5; padding: 15px; border-radius: 4px; border-left: 4px solid #9C27B0;">
name: "Traitement de donn√©es"
description: "Import et validation"
shared_arguments:
  - code: "DATABASE_FILE"
    name: "Base de donn√©es"
    required: 1
    type: "file"
  
  - code: "LOG_LEVEL"
    name: "Niveau de log"
    required: 0
    type: "text"
    default: "INFO"

commands:
  - name: "Import"
    description: "Importe les donn√©es"
    command: "import.exe --db {DATABASE_FILE} --log {LOG_LEVEL}"
    arguments: []
  
  - name: "Validation"
    description: "Valide les donn√©es"
    command: "validate.exe --db {DATABASE_FILE} --log {LOG_LEVEL}"
    arguments: []
        </pre>
        
        <h3>Mapping vers les commandes</h3>
        <p>
        Si les codes ne correspondent pas exactement, utilisez le champ <code>shared_argument_mapping</code> :
        </p>
        <pre style="background-color: #f5f5f5; padding: 15px; border-radius: 4px; border-left: 4px solid #9C27B0;">
shared_arguments:
  - code: "INPUT_FILE"
    name: "Fichier source"
    required: 1
    type: "file"

commands:
  - name: "Import"
    command: "import.exe --source {SOURCE_FILE}"
    arguments: []
    shared_argument_mapping:
      INPUT_FILE: "SOURCE_FILE"
  
  - name: "Backup"
    command: "backup.exe --file {FILE_PATH}"
    arguments: []
    shared_argument_mapping:
      INPUT_FILE: "FILE_PATH"
        </pre>
        
        <h3>Combinaison avec arguments locaux</h3>
        <p>Une commande peut avoir √† la fois des arguments partag√©s et des arguments locaux :</p>
        <pre style="background-color: #f5f5f5; padding: 15px; border-radius: 4px; border-left: 4px solid #9C27B0;">
shared_arguments:
  - code: "DATABASE"
    name: "Base de donn√©es"
    required: 1
    type: "file"

commands:
  - name: "Export"
    command: "export.exe --db {DATABASE} --format {FORMAT}"
    arguments:
      - code: "FORMAT"
        name: "Format de sortie"
        required: 1
        type: "text"
        default: "CSV"
        </pre>
        """
        self.shared_text.setHtml(content)
    
    def _populate_validation(self):
        """Remplit l'onglet Validation."""
        content = """
        <h2>Validation des arguments</h2>
        
        <h3>Types de validation disponibles</h3>
        
        <h4>1. Pattern (expression r√©guli√®re)</h4>
        <p>Valide que la valeur correspond √† un motif regex.</p>
        <pre style="background-color: #f5f5f5; padding: 15px; border-radius: 4px; border-left: 4px solid #F44336;">
arguments:
  - code: "EMAIL"
    name: "Adresse email"
    required: 1
    validation:
      pattern: "^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\\.[a-zA-Z]{2,}$"
      message: "Format d'email invalide"
        </pre>
        
        <h4>2. Min / Max (longueur)</h4>
        <p>Valide la longueur minimale et/ou maximale d'une cha√Æne.</p>
        <pre style="background-color: #f5f5f5; padding: 15px; border-radius: 4px; border-left: 4px solid #F44336;">
arguments:
  - code: "PASSWORD"
    name: "Mot de passe"
    required: 1
    validation:
      min: 8
      max: 32
      message: "Le mot de passe doit contenir entre 8 et 32 caract√®res"
        </pre>
        
        <h4>3. Allowed values (valeurs autoris√©es)</h4>
        <p>Limite les valeurs possibles √† une liste pr√©d√©finie.</p>
        <pre style="background-color: #f5f5f5; padding: 15px; border-radius: 4px; border-left: 4px solid #F44336;">
arguments:
  - code: "LOG_LEVEL"
    name: "Niveau de log"
    required: 1
    validation:
      allowed_values: ["DEBUG", "INFO", "WARNING", "ERROR"]
      message: "Valeur autoris√©e : DEBUG, INFO, WARNING, ERROR"
        </pre>
        
        <h4>4. File exists (fichier existe)</h4>
        <p>V√©rifie que le fichier sp√©cifi√© existe sur le disque.</p>
        <pre style="background-color: #f5f5f5; padding: 15px; border-radius: 4px; border-left: 4px solid #F44336;">
arguments:
  - code: "CONFIG_FILE"
    name: "Fichier de configuration"
    required: 1
    type: "file"
    validation:
      file_exists: true
      message: "Le fichier doit exister"
        </pre>
        
        <h3>Combinaison de validations</h3>
        <p>Vous pouvez combiner plusieurs r√®gles de validation :</p>
        <pre style="background-color: #f5f5f5; padding: 15px; border-radius: 4px; border-left: 4px solid #F44336;">
arguments:
  - code: "USERNAME"
    name: "Nom d'utilisateur"
    required: 1
    validation:
      min: 3
      max: 20
      pattern: "^[a-zA-Z0-9_]+$"
      message: "3-20 caract√®res alphanum√©riques et underscore uniquement"
        </pre>
        
        <h3>Messages d'erreur</h3>
        <p>
        Le champ <code>message</code> est optionnel mais recommand√©. Il affiche un message 
        clair √† l'utilisateur en cas d'erreur de validation.
        </p>
        
        <h3>Validation automatique</h3>
        <p>
        La validation est effectu√©e automatiquement :
        </p>
        <ul>
            <li>Avant l'ex√©cution de la t√¢che</li>
            <li>Une bo√Æte de dialogue affiche toutes les erreurs</li>
            <li>L'ex√©cution est bloqu√©e tant qu'il y a des erreurs</li>
        </ul>
        """
        self.validation_text.setHtml(content)
    
    def _populate_examples(self):
        """Remplit l'onglet Exemples Complets."""
        content = """
        <h2>Exemples complets</h2>
        
        <h3>Exemple 1 : T√¢che simple</h3>
        <pre style="background-color: #f5f5f5; padding: 15px; border-radius: 4px; border-left: 4px solid #00BCD4;">
name: "Sauvegarde de fichiers"
description: "Copie des fichiers vers un dossier de backup"
commands:
  - name: "Cr√©er le dossier"
    description: "Cr√©e le dossier de destination"
    command: "mkdir {BACKUP_DIR}"
    arguments:
      - code: "BACKUP_DIR"
        name: "Dossier de backup"
        required: 1
        type: "folder"
  
  - name: "Copier les fichiers"
    description: "Copie les fichiers"
    command: "xcopy {SOURCE} {BACKUP_DIR} /E /I"
    arguments:
      - code: "SOURCE"
        name: "Dossier source"
        required: 1
        type: "folder"
      - code: "BACKUP_DIR"
        name: "Dossier de backup"
        required: 1
        type: "folder"
        </pre>
        
        <h3>Exemple 2 : Avec arguments partag√©s</h3>
        <pre style="background-color: #f5f5f5; padding: 15px; border-radius: 4px; border-left: 4px solid #00BCD4;">
name: "Pipeline de traitement"
description: "Import, validation et export de donn√©es"
shared_arguments:
  - code: "DATABASE"
    name: "Base de donn√©es"
    required: 1
    type: "file"
    validation:
      file_exists: true
      message: "La base de donn√©es doit exister"
  
  - code: "VERBOSE"
    name: "Mode verbeux"
    required: 0
    type: "text"
    default: "false"
    validation:
      allowed_values: ["true", "false"]
      message: "Valeur : true ou false"

commands:
  - name: "Import"
    description: "Importe les donn√©es CSV"
    command: "import.exe --db {DATABASE} --source {CSV_FILE} --verbose {VERBOSE}"
    arguments:
      - code: "CSV_FILE"
        name: "Fichier CSV"
        required: 1
        type: "file"
  
  - name: "Validation"
    description: "Valide les donn√©es import√©es"
    command: "validate.exe --db {DATABASE} --verbose {VERBOSE}"
    arguments: []
  
  - name: "Export"
    description: "Export vers Excel"
    command: "export.exe --db {DATABASE} --output {OUTPUT} --verbose {VERBOSE}"
    arguments:
      - code: "OUTPUT"
        name: "Fichier de sortie"
        required: 1
        type: "file"
        default: "output.xlsx"
        </pre>
        
        <h3>Exemple 3 : Avec validation avanc√©e</h3>
        <pre style="background-color: #f5f5f5; padding: 15px; border-radius: 4px; border-left: 4px solid #00BCD4;">
name: "Configuration serveur"
description: "Configure un serveur avec validation"
commands:
  - name: "Configuration"
    description: "Configure le serveur"
    command: "configure.exe --host {HOST} --port {PORT} --user {USER} --pass {PASS}"
    arguments:
      - code: "HOST"
        name: "Adresse du serveur"
        required: 1
        type: "text"
        validation:
          pattern: "^([0-9]{1,3}\\.){3}[0-9]{1,3}$|^[a-zA-Z0-9.-]+$"
          message: "Adresse IP ou nom de domaine valide requis"
      
      - code: "PORT"
        name: "Port"
        required: 1
        type: "text"
        default: "8080"
        validation:
          pattern: "^[0-9]{1,5}$"
          message: "Port entre 1 et 65535"
      
      - code: "USER"
        name: "Nom d'utilisateur"
        required: 1
        type: "text"
        validation:
          min: 3
          max: 20
          pattern: "^[a-zA-Z0-9_]+$"
          message: "3-20 caract√®res alphanum√©riques"
      
      - code: "PASS"
        name: "Mot de passe"
        required: 1
        type: "text"
        validation:
          min: 8
          message: "Minimum 8 caract√®res"
        </pre>
        
        <h3>Exemple 4 : Mapping d'arguments partag√©s</h3>
        <pre style="background-color: #f5f5f5; padding: 15px; border-radius: 4px; border-left: 4px solid #00BCD4;">
name: "Traitement multi-outils"
description: "Utilise plusieurs outils avec des noms d'arguments diff√©rents"
shared_arguments:
  - code: "INPUT_FILE"
    name: "Fichier d'entr√©e"
    required: 1
    type: "file"
  
  - code: "OUTPUT_DIR"
    name: "Dossier de sortie"
    required: 1
    type: "folder"

commands:
  - name: "Conversion"
    description: "Convertit le fichier"
    command: "convert.exe --source {SRC} --destination {DST}"
    arguments: []
    shared_argument_mapping:
      INPUT_FILE: "SRC"
      OUTPUT_DIR: "DST"
  
  - name: "Validation"
    description: "Valide le r√©sultat"
    command: "validate.exe --file {FILE} --outdir {OUT}"
    arguments: []
    shared_argument_mapping:
      INPUT_FILE: "FILE"
      OUTPUT_DIR: "OUT"
        </pre>
        
        <h3>üí° Conseils</h3>
        <ul>
            <li>Utilisez des noms explicites pour les arguments</li>
            <li>Ajoutez des descriptions claires</li>
            <li>D√©finissez des valeurs par d√©faut quand c'est pertinent</li>
            <li>Utilisez la validation pour √©viter les erreurs</li>
            <li>Privil√©giez les arguments partag√©s pour √©viter la duplication</li>
            <li>Testez vos fichiers YAML avant de les d√©ployer</li>
        </ul>
        """
        self.examples_text.setHtml(content)
